### PR TITLE
Build system

### PR DESCRIPTION
**Pull request by realamirhe**
Please double-check the process locally and release on the RC tag in npm

### CODE REVIEW

**Comment by realamirhe** [comment]
The problem of the ci seems to be because of GitHub actions node_modules cache.

It is totally okay locally, and I believe if we switch to yarn it would be okay too
but in case we want to stick to npm, Do you have any idea how to fix it @chintan9 ?
https://stackoverflow.com/a/49505612/10321531

**Comment by chintan9** [comment]
> The problem of the ci seems to be because of GitHub actions node_modules cache.
> 
> It is totally okay locally, and I believe if we switch to yarn it would be okay too but in case we want to stick to npm, Do you have any idea how to fix it @chintan9 ? https://stackoverflow.com/a/49505612/10321531

```
(!) Plugin typescript: @rollup/plugin-typescript TS2322: Type '(_: any, params: PlyrConfigurationProps) => PlyrJS' is not assignable to type 'Instantiate<Plyr, HTMLVideoElement, PlyrConfigurationProps>'.
  Types of parameters 'params' and 'params' are incompatible.
    Type 'PlyrConfigurationProps | undefined' is not assignable to type 'PlyrConfigurationProps'.
      Type 'undefined' is not assignable to type 'PlyrConfigurationProps'.
src/index.tsx: (74:7)

74       instantiate,
         ~~~~~~~~~~~

  node_modules/react-aptor/use-aptor.d.ts:8:5
    8     instantiate: Instantiate<TInstance, TElement, TParams>;
          ~~~~~~~~~~~
    The expected type comes from property 'instantiate' which is declared here on type 'AptorConfiguration<Plyr, HTMLVideoElement, PlyrConfigurationProps>'
    
    ```

**Comment by chintan9** [comment]
can we use SWC instead of esbuild?

**Comment by realamirhe** [comment]
> can we use SWC instead of esbuild?
> 

I tried that, but I took too much time to install and got failed at the end.

If you can install
`@swc/core` and `@swc/jest` and update the package-lock.json;
I will do the setup on my own

**Comment by realamirhe** [comment]
> > The problem of the ci seems to be because of GitHub actions node_modules cache.
> > 
> > It is totally okay locally, and I believe if we switch to yarn it would be okay too but in case we want to stick to npm, Do you have any idea how to fix it @chintan9 ? https://stackoverflow.com/a/49505612/10321531
> 
> ```
> (!) Plugin typescript: @rollup/plugin-typescript TS2322: Type '(_: any, params: PlyrConfigurationProps) => PlyrJS' is not assignable to type 'Instantiate<Plyr, HTMLVideoElement, PlyrConfigurationProps>'.
>   Types of parameters 'params' and 'params' are incompatible.
>     Type 'PlyrConfigurationProps | undefined' is not assignable to type 'PlyrConfigurationProps'.
>       Type 'undefined' is not assignable to type 'PlyrConfigurationProps'.
> src/index.tsx: (74:7)
> 
> 74       instantiate,
>          ~~~~~~~~~~~
> 
>   node_modules/react-aptor/use-aptor.d.ts:8:5
>     8     instantiate: Instantiate<TInstance, TElement, TParams>;
>           ~~~~~~~~~~~
>     The expected type comes from property 'instantiate' which is declared here on type 'AptorConfiguration<Plyr, HTMLVideoElement, PlyrConfigurationProps>'
>     
>     ```



I'm pretty sure, I fixed this😅,
Okay let me take another look to it

**Comment by realamirhe** [comment]
Hey @chintan9, I fixed the type `Instantiate` Bug, it was due to a type limitation, now we have 3 warnings but there is a completely valid statement, and have a plan to fix them soon.

```bash
/plyr-react/src/index.tsx
  37:42  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
  38:7   warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
  38:37  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
```

The problem with the `plyr`
> (!) Plugin typescript: @rollup/plugin-typescript TS2307: Cannot find module 'plyr' or its corresponding type declarations.

Seems to exist, because of the minimum version of our npm in the pacakge.json engine
I was using the `node: v16.14.0` and `npm: 8.3.1` so I didn't have any issues.

Here are some links which can be helpful
- [stack overflow](https://stackoverflow.com/a/35207983/10321531)
- [install peer deps](https://github.com/npm/rfcs/blob/main/implemented/0025-install-peer-deps.md)
- [Presenting v7.0.0 of the npm CLI](https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/)

**Comment by realamirhe** [comment]
Now we have another problem with the react testing library, I remember I solved that on my side (locally) via `--legacy-peer-deps` which is not a good option.

I'm going to fix the react testing library react peer deps after Monday.
And we might be able to publish the latest soon
@chintan9

**Comment by realamirhe** [comment]
Fixed the conflicts and build errors, all kudos to dear gitpod :)

**Comment by realamirhe** [comment]
@chintan9 although I love to pass all these commit to master on my name?
I changed the package-lock.json and it faced conflicts, so I was forced to merge master into the build-system, so it is up to you what to do, but it seems the first is more professional action.

1. squash and merge 
image

2. Merge pull request

File: rollup.config.js [hunk 1]
```diff
@@ -1,34 +1,137 @@
...
-import babel from '@rollup/plugin-babel'
+import path from "path";
+import babelPlugin from "@rollup/plugin-babel";
+import resolve from "@rollup/plugin-node-resolve";
+import replace from "@rollup/plugin-replace";
+import typescript from "@rollup/plugin-typescript";
+import copy from "rollup-plugin-copy";
+import esbuild from "rollup-plugin-esbuild";
+import { terser } from "rollup-plugin-terser";
+import copy from "rollup-plugin-copy";
```
[the remark this answers is not in the document]
Comment (realamirhe) [reply rollup.config.js]: @chintan9 this is not a relevant case, you can double-check the `rollup.config.js` file in the changes section

**Review by chintan9 — APPROVED** [review]

**Pull request merged**